File: sdk/featureflags_sdk_test.py
import sys
import types
import unittest
from unittest.mock import patch

if "requests" not in sys.modules:
    fake_requests = types.SimpleNamespace(get=None, post=None)
    sys.modules["requests"] = fake_requests

from featureflags_sdk import FeatureFlagsClient


class FakeResponse:
    def __init__(self, payload=None, ok=True):
        self._payload = payload or {}
        self.ok = ok

    def json(self):
        return self._payload

    def raise_for_status(self):
        return None


class FeatureFlagsClientTests(unittest.TestCase):
    def test_record_conversion_uses_bearer_token(self):
        client = FeatureFlagsClient("http://localhost:8080", auth_token="secret-token")

        with patch("featureflags_sdk.requests.post", return_value=FakeResponse()) as post:
            client.record_conversion("button-color", "green")

        post.assert_called_once()
        _, kwargs = post.call_args
        self.assertEqual(kwargs["headers"], {"Authorization": "Bearer secret-token"})
        self.assertEqual(kwargs["params"], {"variant": "green"})

    def test_report_circuit_breaker_uses_bearer_token(self):
        client = FeatureFlagsClient("http://localhost:8080", auth_token="secret-token")

        with patch("featureflags_sdk.requests.post", return_value=FakeResponse()) as post:
            client.report_circuit_breaker("/demo/action", False, 320)

        post.assert_called_once()
        _, kwargs = post.call_args
        self.assertEqual(kwargs["headers"], {"Authorization": "Bearer secret-token"})
        self.assertEqual(
            kwargs["json"],
            {"route": "/demo/action", "success": False, "latencyMs": 320},
        )

    def test_check_circuit_breaker_returns_allowed_and_state(self):
        client = FeatureFlagsClient("http://localhost:8080")

        with patch(
            "featureflags_sdk.requests.post",
            return_value=FakeResponse({"allowed": False, "state": "half-open"}),
        ):
            allowed, state = client.check_circuit_breaker("/demo/action")

        self.assertFalse(allowed)
        self.assertEqual(state, "half-open")


if __name__ == "__main__":
    unittest.main()

File: sdk/featureflags_sdk.test.js
const assert = require('node:assert/strict');
const Module = require('node:module');

let fetchImpl = async () => ({ ok: true, json: async () => ({}) });

const originalLoad = Module._load;
Module._load = function patchedLoad(request, parent, isMain) {
  if (request === 'node-fetch') {
    return (...args) => fetchImpl(...args);
  }
  if (request === 'eventsource') {
    return class FakeEventSource {};
  }
  return originalLoad(request, parent, isMain);
};

const FeatureFlagsClient = require('./featureflags_sdk');
Module._load = originalLoad;

async function testRecordConversionUsesBearerToken() {
  const calls = [];
  fetchImpl = async (url, options = {}) => {
    calls.push({ url, options });
    return { ok: true, json: async () => ({}) };
  };

  const client = new FeatureFlagsClient('http://localhost:8080', 'dev', 'secret-token');
  await client.recordConversion('button-color', 'green');

  assert.equal(calls.length, 1);
  assert.equal(
    calls[0].url,
    'http://localhost:8080/experiment/button-color/convert?variant=green'
  );
  assert.equal(calls[0].options.method, 'POST');
  assert.equal(calls[0].options.headers.Authorization, 'Bearer secret-token');
}

async function testReportCircuitBreakerUsesBearerToken() {
  const calls = [];
  fetchImpl = async (url, options = {}) => {
    calls.push({ url, options });
    return { ok: true, json: async () => ({}) };
  };

  const client = new FeatureFlagsClient('http://localhost:8080', 'dev', 'secret-token');
  await client.reportCircuitBreakerResult('/demo/action', false, 320);

  assert.equal(calls.length, 1);
  assert.equal(calls[0].url, 'http://localhost:8080/circuitbreaker/report');
  assert.equal(calls[0].options.method, 'POST');
  assert.equal(calls[0].options.headers.Authorization, 'Bearer secret-token');
  assert.equal(
    calls[0].options.body,
    JSON.stringify({ route: '/demo/action', success: false, latencyMs: 320 })
  );
}

async function testCheckCircuitBreakerParsesResponse() {
  fetchImpl = async () => ({
    ok: true,
    json: async () => ({ allowed: false, state: 'half-open' }),
  });

  const client = new FeatureFlagsClient('http://localhost:8080');
  const result = await client.checkCircuitBreaker('/demo/action');

  assert.deepEqual(result, { allowed: false, state: 'half-open' });
}

async function main() {
  await testRecordConversionUsesBearerToken();
  await testReportCircuitBreakerUsesBearerToken();
  await testCheckCircuitBreakerParsesResponse();
  console.log('featureflags_sdk.test.js: PASS');
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});